pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,198 @@
+	MarkedForDelete           = true
+	EnforceRevisionCheckParam = false
+	ResourceTypeSubnet        = common.ResourceTypeSubnet
+	NewConverter              = common.NewConverter
+	// Default static ip-pool under Subnet.
+	ipPoolID = "static-ipv4-default"
+)
+
+type SubnetService struct {
+	common.Service
+	SubnetStore *SubnetStore
+}
+
+// SubnetParameters stores parameters to CRUD Subnet object
+type SubnetParameters struct {
+	OrgID     string
+	ProjectID string
+	VPCID     string
+}
+
+var subnetService *SubnetService
+var lock = &sync.Mutex{}
+
+// GetSubnetService get singleton SubnetService instance, subnet/subnetset controller share the same instance.
+func GetSubnetService(service common.Service) *SubnetService {
+	if subnetService == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if subnetService == nil {
+			var err error
+			if subnetService, err = InitializeSubnetService(service); err != nil {
+				log.Error(err, "failed to initialize subnet commonService")
+				os.Exit(1)
+			}
+		}
+	}
+	return subnetService
+}
+
+// InitializeSubnetService initialize Subnet service.
+func InitializeSubnetService(service common.Service) (*SubnetService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+	subnetService := &SubnetService{
+		Service: service,
+		SubnetStore: &SubnetStore{
+			ResourceStore: common.ResourceStore{
+				Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeSubnetCRUID: indexFunc}),
+				BindingType: model.VpcSubnetBindingType(),
+			},
+		},
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+func (service *SubnetService) CreateOrUpdateSubnet(obj *v1alpha1.Subnet, orgID, projectID, vpcID string) error {
+	nsxSubnet, err := service.buildSubnet(obj)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return err
+	}
+	existingSubnet := service.SubnetStore.GetByKey(*nsxSubnet.Id)
+	changed := false
+	if existingSubnet == nil {
+		changed = true
+	} else {
+		changed = common.CompareResource(SubnetToComparable(existingSubnet), SubnetToComparable(nsxSubnet))
+	}
+	if !changed {
+		log.Info("subnet not changed, skip updating", "subnet.Id", *nsxSubnet.Id)
+		return nil
+	}
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, orgID, projectID, vpcID)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return err
+	}
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(orgID, projectID, vpcID, *nsxSubnet.Id); err != nil {
+		return err
```

Comment:
For VPC Subnet, after create it via Policy API, do we need to check its realization state? Is there possibility that the Policy intent succeeds, but the Subnet is realized failed?

Reply:
thanks, let me try this.

Reply:
Already added a service to check realization status.